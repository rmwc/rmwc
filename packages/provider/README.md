# Provider

> A component that provides global configuration for RMWC.

-   Module __@rmwc/provider__

You can wrap your top level App component with RMWC provider to set global configuration options. Just pass the options in as props.

```js

  import React from 'react';

  import \* as ReactDOM from 'react-dom';

  import App from './App'; // your main app component

  import { RMWCProvider } from '@rmwc/provider';

  // This example disables ripples globally by default

  ReactDOM.render(

    <RMWCProvider

      // Globally disable ripples

      ripple\={false}

      // Global options for icons

      // Takes the same options as the icon component

      icon\={{

        basename: 'material-icons'

      }}

      // Global options for typography

      // allows mapping of a defaultTag or specific classes

      // See the Typography docs for more info

      typography\={{

        defaultTag: 'div',

        headline1: 'h1'

      }}

      // Global options for tooltips

      // Takes most of the options for tooltips

      // See the Tooltip docs for more info

      tooltip\={{

        align: 'right'

      }}

    \>

      <App />

    </RMWCProvider\>,

    document.getElementById('root'),

  );

```

## RMWCProvider