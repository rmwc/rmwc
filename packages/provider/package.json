{
  "name": "@rmwc/provider",
  "version": "14.0.1",
  "description": "RMWC Provider component",
  "main": "./index.js",
  "types": "./index.d.ts",
  "exports": {
    ".": {
      "import": "./index.mjs",
      "types": "./index.d.ts",
      "require": "./index.js"
    }
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rmwc/rmwc.git"
  },
  "keywords": [
    "rmwc"
  ],
  "author": "rmwc",
  "private": false,
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/rmwc/rmwc/issues"
  },
  "homepage": "https://github.com/rmwc/rmwc/tree/master/packages/provider",
  "peerDependencies": {
    "react": ">=16.8.x",
    "react-dom": ">=16.8.x"
  },
  "dependencies": {
    "@rmwc/base": "workspace:*",
    "@rmwc/types": "workspace:*"
  }
}
