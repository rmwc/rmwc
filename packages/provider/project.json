{
  "name": "provider",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/provider/src",
  "projectType": "library",
  "tags": [
    "publishable",
    "type:ui-component",
    "readme"
  ],
  "targets": {
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "packages/provider/**/*.{ts,tsx,js,jsx}"
        ]
      }
    },
    "build": {
      "executor": "@nx/vite:build",
      "dependsOn": [{
        "projects": "readme",
        "target": "generate-markdown"
      }],
      "outputs": [
        "{options.outputPath}"
      ],
      "defaultConfiguration": "production",
      "options": {
        "outputPath": "dist/packages/provider"
      },
      "configurations": {
        "development": {
          "mode": "development"
        },
        "production": {
          "mode": "production"
        }
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": [
        "{options.reportsDirectory}"
      ],
      "options": {
        "passWithNoTests": true,
        "reportsDirectory": "../../coverage/packages/provider"
      }
    },
    "npm-deploy": {
      "executor": "ngx-deploy-npm:deploy",
      "options": {
        "access": "public",
        "buildTarget": "production"
      }
    }
  }
}
