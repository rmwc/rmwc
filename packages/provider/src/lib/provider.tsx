import React from 'react';
import * as RMWC from '@rmwc/types';

type TooltipActivationT = 'hover' | 'click' | 'focus';

type RCTooltipAlignT =
  | 'left'
  | 'right'
  | 'top'
  | 'bottom'
  | 'topLeft'
  | 'topRight'
  | 'bottomLeft'
  | 'bottomRight';

type TooltipAlignT =
  | 'start'
  | 'center'
  | 'end'
  | 'above'
  | 'below'
  | 'startAbove'
  | 'startBelow'
  | 'centerAbove'
  | 'centerBelow'
  | 'endAbove'
  | 'endBelow';

export type TooltipOptions = {
  /** How to align the tooltip. This affects both RCTooltip and Tooltip, but only if the given alignment is supported by the component. Defaults to `top`. */
  align?: RCTooltipAlignT | TooltipAlignT;
  /** Activate the tooltip through one or more interactions. Defaults to `['hover', 'focus']`. */
  activateOn?: TooltipActivationT | TooltipActivationT[];
  /** Whether or not to show an arrow on the Tooltip. Only supported by RCTooltip. Defaults to `false`. */
  showArrow?: boolean;
  /** Delay in milliseconds before showing the tooltip when interacting via touch or mouse. */
  enterDelay?: number;
  /** Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. */
  leaveDelay?: number;
};

type TypographyOptions = {
  defaultTag?: string | React.ComponentType<any>;
  headline1?: string | React.ComponentType<any>;
  headline2?: string | React.ComponentType<any>;
  headline3?: string | React.ComponentType<any>;
  headline4?: string | React.ComponentType<any>;
  headline5?: string | React.ComponentType<any>;
  headline6?: string | React.ComponentType<any>;
  subtitle1?: string | React.ComponentType<any>;
  subtitle2?: string | React.ComponentType<any>;
  body1?: string | React.ComponentType<any>;
  body2?: string | React.ComponentType<any>;
  caption?: string | React.ComponentType<any>;
  button?: string | React.ComponentType<any>;
  overline?: string | React.ComponentType<any>;
};

/** A provider for setting global options in RMWC. */
export interface RMWCProviderProps {
  /** Enable / Disable interaction ripples globally */
  ripple?: boolean;
  /** Global options for icons */
  icon?: Partial<RMWC.IconOptions>;
  /** Global tooltip options */
  tooltip?: Partial<TooltipOptions>;
  /** Global typography options */
  typography?: Partial<TypographyOptions>;
  /** Children to render */
  children?: React.ReactNode;
}

// Default provider options
const providerDefaults: RMWCProviderProps = {
  ripple: true,
  tooltip: {
    align: 'top',
    showArrow: false,
    activateOn: ['hover', 'focus'],
    enterDelay: 0,
    leaveDelay: 0
  },
  typography: {},
  icon: {
    icon: '',
    basename: 'material-icons',
    prefix: '',
    strategy: 'auto',
    render: undefined
  }
};

export const ProviderContext = React.createContext(providerDefaults);

export const useProviderContext = () => React.useContext(ProviderContext);

/** A provider for setting global options in RMWC. */
export const RMWCProvider = ({ children, ...rest }: RMWCProviderProps) => {
  return (
    <ProviderContext.Provider
      value={{
        ...providerDefaults,
        ...rest
      }}
    >
      {children}
    </ProviderContext.Provider>
  );
};
