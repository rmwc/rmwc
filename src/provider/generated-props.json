{
  "npm": {},
  "typescript": {
    "RMWCProviderProps": {
      "documentation": {
        "contents": [
          "<p>A provider for setting global options in RMWC.</p>\n"
        ],
        "contentsRaw": "A provider for setting global options in RMWC.",
        "metadata": {}
      },
      "fileName": "src/provider/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "RMWCProviderProps",
      "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L47",
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>Children to render</p>\n"
            ],
            "contentsRaw": "Children to render",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "children",
          "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L57",
          "type": "React.ReactNode"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global options for icons</p>\n"
            ],
            "contentsRaw": "Global options for icons",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "icon",
          "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L51",
          "type": "Partial<RMWC.IconOptions>"
        },
        {
          "documentation": {
            "contents": [
              "<p>Enable / Disable interaction ripples globally</p>\n"
            ],
            "contentsRaw": "Enable / Disable interaction ripples globally",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "ripple",
          "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L49",
          "type": "undefined | false | true"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global tooltip options</p>\n"
            ],
            "contentsRaw": "Global tooltip options",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "tooltip",
          "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L53",
          "type": "Partial<TooltipOptions>"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global typography options</p>\n"
            ],
            "contentsRaw": "Global typography options",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "typography",
          "sourceUrl": "https://github.com/jamesmfriedman/rmwc/blob/master/src/provider/index.tsx#L55",
          "type": "Partial<TypographyOptions>"
        }
      ]
    }
  }
}
