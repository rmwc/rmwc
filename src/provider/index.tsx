import React from 'react';
import * as RMWC from '@rmwc/types';

type TooltipActivationT = 'hover' | 'click' | 'focus';

type TooltipAlignT =
  | 'left'
  | 'right'
  | 'top'
  | 'bottom'
  | 'topLeft'
  | 'topRight'
  | 'bottomLeft'
  | 'bottomRight';

type TooltipOptions = {
  /** How to align the tooltip. Defaults to `top`. */
  align?: TooltipAlignT;
  /** Activate the tooltip through one or more interactions. Defaults to `['hover', 'focus']`. */
  activateOn?: TooltipActivationT | TooltipActivationT[];
  /** Whether or not to show an arrow on the Tooltip. Defaults to `false`. */
  showArrow?: boolean;
  /** Delay in milliseconds before showing the tooltip when interacting via touch or mouse. */
  enterDelay?: number;
  /** Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. */
  leaveDelay?: number;
};

type TypographyOptions = {
  defaultTag?: string | React.ComponentType<any>;
  headline1?: string | React.ComponentType<any>;
  headline2?: string | React.ComponentType<any>;
  headline3?: string | React.ComponentType<any>;
  headline4?: string | React.ComponentType<any>;
  headline5?: string | React.ComponentType<any>;
  headline6?: string | React.ComponentType<any>;
  subtitle1?: string | React.ComponentType<any>;
  subtitle2?: string | React.ComponentType<any>;
  body1?: string | React.ComponentType<any>;
  body2?: string | React.ComponentType<any>;
  caption?: string | React.ComponentType<any>;
  button?: string | React.ComponentType<any>;
  overline?: string | React.ComponentType<any>;
};

/** A provider for setting global options in RMWC. */
export interface RMWCProviderProps {
  /** Enable / Disable interaction ripples globally */
  ripple?: boolean;
  /** Global options for icons */
  icon?: Partial<RMWC.IconOptions>;
  /** Global tooltip options */
  tooltip?: Partial<TooltipOptions>;
  /** Global typography options */
  typography?: Partial<TypographyOptions>;
  /** Children to render */
  children?: React.ReactNode;
}

// Default provider options
const providerDefaults: RMWCProviderProps = {
  ripple: true,
  tooltip: {
    align: 'top',
    showArrow: false,
    activateOn: ['hover', 'focus'],
    enterDelay: 0,
    leaveDelay: 0
  },
  typography: {},
  icon: {
    icon: '',
    basename: 'material-icons',
    prefix: '',
    strategy: 'auto',
    render: undefined
  }
};

export const ProviderContext = React.createContext(providerDefaults);

export const useProviderContext = () => React.useContext(ProviderContext);

/** A provider for setting global options in RMWC. */
export const RMWCProvider = ({ children, ...rest }: RMWCProviderProps) => {
  return (
    <ProviderContext.Provider
      value={{
        ...providerDefaults,
        ...rest
      }}
    >
      {children}
    </ProviderContext.Provider>
  );
};
