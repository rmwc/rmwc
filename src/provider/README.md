# Provider

A component that provides global configuration for RMWC.

- Module **@rmwc/provider**


You can wrap your top level App component with RMWC provider to set global configuration options. Just pass the options in as props.

```jsx

  `
  import React from 'react';
  import * as ReactDOM from 'react-dom';
  import App from './App'; // your main app component
  import { RMWCProvider } from '@rmwc/provider';

  // This example disables ripples globally by default
  ReactDOM.render(
    <RMWCProvider
      // Globally disable ripples
      ripple={false}
      // Global options for icons
      // Takes the same options as the icon component
      icon={{
        basename: 'material-icons'
      }}
      // Global options for typography
      // allows mapping of a defaultTag or specific classes
      // See the Typography docs for more info
      typography={{
        defaultTag: 'div',
        headline1: 'h1'
      }}
      // Global options for tooltips
      // Takes most of the options for tooltips
      // See the Tooltip docs for more info
      tooltip={{
        align: 'right'
      }}
    >
      <App />
    </RMWCProvider>,
    document.getElementById('root'),
  );
`

```

## RMWCProvider
A provider for setting global options in RMWC.

### Props

| Name | Type | Description |
|------|------|-------------|
| `children` | `React.ReactNode` | Children to render |
| `icon` | `Partial<IconOptions>` | Global options for icons |
| `ripple` | `undefined \| false \| true` | Enable / Disable interaction ripples globally |
| `tooltip` | `Partial<TooltipOptions>` | Global tooltip options |
| `typography` | `Partial<TypographyOptions>` | Global typography options |


