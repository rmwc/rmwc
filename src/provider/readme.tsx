import React from 'react';

import { Docs, DocsExample, DocProps, DocsP } from '@doc-utils';
import propsSrc from './generated-props.json';
import examples from './generated-examples.json';

import { RMWCProvider } from '.';

export default function Readme() {
  return (
    <Docs
      title="Provider"
      lead="A component that provides global configuration for RMWC."
      module="@rmwc/provider"
      styles={[]}
      examples={examples}
    >
      <DocsP>
        You can wrap your top level App component with RMWC provider to set
        global configuration options. Just pass the options in as props.
      </DocsP>

      <DocsExample codeOnly>
        {
          /* jsx */ `
          import React from 'react';
          import * as ReactDOM from 'react-dom';
          import App from './App'; // your main app component
          import { RMWCProvider } from '@rmwc/provider';

          // This example disables ripples globally by default
          ReactDOM.render(
            <RMWCProvider
              // Globally disable ripples
              ripple={false}
              // Global options for icons
              // Takes the same options as the icon component
              icon={{
                basename: 'material-icons'
              }}
              // Global options for typography
              // allows mapping of a defaultTag or specific classes
              // See the Typography docs for more info
              typography={{
                defaultTag: 'div',
                headline1: 'h1'
              }}
              // Global options for tooltips
              // Takes most of the options for tooltips
              // See the Tooltip docs for more info
              tooltip={{
                align: 'right'
              }}
            >
              <App />
            </RMWCProvider>,
            document.getElementById('root'),
          );
        `
        }
      </DocsExample>

      <DocProps
        src={propsSrc}
        components={[{ displayName: 'RMWCProvider', component: RMWCProvider }]}
      />
    </Docs>
  );
}
