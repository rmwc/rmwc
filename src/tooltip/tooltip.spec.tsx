import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Tooltip } from './';
import { RMWCProvider } from '../provider';

describe('Tooltip', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <Tooltip content="tooltip">
        <span>test</span>
      </Tooltip>
    );
    userEvent.hover(screen.getByText('test'));
    expect(screen.getByText('tooltip')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('activateOn', () => {
    render(
      <Tooltip content="tooltip" activateOn="click">
        <span>test</span>
      </Tooltip>
    );
  });

  it('showArrow', () => {
    const { asFragment } = render(
      <Tooltip content="tooltip" showArrow>
        <span>test</span>
      </Tooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('className', () => {
    render(
      <Tooltip content="tooltip" className="my-custom-classname">
        <span>test</span>
      </Tooltip>
    );
  });

  it('enterDelay', () => {
    render(
      <Tooltip content="tooltip" enterDelay={1000}>
        <span>test</span>
      </Tooltip>
    );
  });

  it('leaveDelay', () => {
    render(
      <Tooltip content="tooltip" leaveDelay={1000}>
        <span>test</span>
      </Tooltip>
    );
  });

  it('align', () => {
    const { asFragment } = render(
      <Tooltip content="tooltip" align="bottom">
        <span>test</span>
      </Tooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('works with provider', () => {
    render(
      <RMWCProvider tooltip={{}}>
        <Tooltip content="tooltip">
          <span>test</span>
        </Tooltip>
      </RMWCProvider>
    );
  });
});
