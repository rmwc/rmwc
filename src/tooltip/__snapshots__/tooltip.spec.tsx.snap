// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip align 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="mdc-tooltip"
    id="tooltip-id"
    role="tooltip"
  >
    <div
      class="mdc-tooltip__surface mdc-tooltip__surface-animation"
    >
      tooltip
    </div>
  </div>
  <span
    aria-describedby="tooltip-id"
    class=""
    data-tooltip-id="tooltip-id"
  >
    test
  </span>
</DocumentFragment>
`;

exports[`Tooltip renders 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="mdc-tooltip mdc-tooltip--shown mdc-tooltip--showing"
    data-mdc-tooltip-persistent="true"
    id="tooltip-id"
    role="tooltip"
    style="top: 4px; left: 0px;"
  >
    <div
      class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      style="transform-origin: center top;"
    >
      tooltip
    </div>
  </div>
  <span
    aria-describedby="tooltip-id"
    class=""
    data-tooltip-id="tooltip-id"
  >
    test
  </span>
</DocumentFragment>
`;
